plugins/lookup/etcd3.py

```diff
@@ -0,0 +1,223 @@
+        msg: "{{ lookup('community.general.etcd3', 'foo/bar', cert_cert='/etc/ssl/etcd/client.pem', cert_key='/etc/ssl/etcd/client.key') }}"
+'''
+
+RETURN = '''
+    _raw:
+        description:
+        - List of keys and associated values.
+        type: list
+        elements: complex
+'''
+
+import re
+
+from ansible.plugins.lookup import LookupBase
+from ansible.utils.display import Display
+from ansible.module_utils.basic import missing_required_lib
+from ansible.module_utils._text import to_native
+from ansible.plugins.lookup import LookupBase
+from ansible.errors import AnsibleError, AnsibleLookupError
+
+try:
+    import etcd3
+    HAS_ETCD = True
+except ImportError:
+    HAS_ETCD = False
+
+display = Display()
+
+etcd3_cnx_opts = (
+    'host',
+    'port',
+    'ca_cert',
+    'cert_key',
+    'cert_cert',
+    'timeout',
+    'user',
+    'password',
+    # 'grpc_options' Etcd3Client() option currently not supported by lookup module (maybe in future ?)
+)
+
+
+def etcd3_client(client_params):
+    try:
+        etcd = etcd3.client(**client_params)
+        etcd.status()
+    except Exception as exp:
+        raise AnsibleLookupError('Cannot connect to etcd cluster: %s' % (to_native(exp)))
+    return etcd
+
+
+class LookupModule(LookupBase):
+
+    def run(self, terms, variables, **kwargs):
+
+        self.set_options(var_options=variables, direct=kwargs)
+
+        if not HAS_ETCD:
+            display.error(missing_required_lib('etcd3'))
+            return None
+
+        # create the etcd3 connection parameters dict to pass to etcd3 class
+        client_params = {}
+
+        # 'endpoints' option is mutually exclusive with 'host' and 'port'
+        if 'endpoints' in kwargs.keys() and 'host' in kwargs.keys() or 'port' in kwargs.keys():
+            raise AnsibleError("etcd3 lookup option 'endpoints' is mutually exclusive with 'host' and 'port' options !")
+
+        # etcd3 class expects host and port as connection parameters, so endpoints
+        # must be mangled a bit to fit in this scheme.
+        # so here we use a regex to extract server and port
+        match = re.compile(
+            r'^(https?://)?(?P<host>(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})|([-_\d\w\.]+))(:(?P<port>\d{1,5}))?/?$'
+        ).match(self.get_option('endpoints'))
+        if match:
+            if match.group('host'):
+                client_params['host'] = match.group('host')
+                display.verbose("etcd3 option 'host' overriden by 'endpoint' host value: %s" % client_params['host'])
+            if match.group('port'):
+                client_params['port'] = match.group('port')
+                display.verbose("etcd3 option 'port' overriden by 'endpoint' port value: %s" % client_params['port'])
+
+        for opt in etcd3_cnx_opts:
+            if self.get_option(opt):
+                client_params[opt] = self.get_option(opt)
+
+        # connect to etcd3 server
+        display.verbose("etcd3 connection parameters: %s" % client_params)
```

Comment:
This potentially dumps the password to the log.
```suggestion
```

Reply:
You're right, I noted this somewhere but I forgot to address this.
As I consider it is important to print this for troubleshooting connection issues, I propose this:
```python
        cnx_log = dict(client_params)
        if 'password' in cnx_log.keys():
            cnx_log.['password'] = '<obfuscated password>'
        display.verbose("etcd3 connection parameters: %s" % cnx_log)
 ```

Reply:
That should be fine!